azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java

```diff
@@ -70,17 +70,22 @@ void setup(final ExecutableFlow flow) {
       // First get the ProjectVersion
       final ProjectVersion projectVersion = getProjectVersion(flow);
 
-      // Setup the project
-      setupProject(projectVersion);
-
-      // Create the execution directory
-      execDir = createExecDir(flow);
-
-      // Create the symlinks from the project
-      copyCreateHardlinkDirectory(projectVersion.getInstalledDir(), execDir);
+      // Synchronized on {@code projectVersion} to prevent
+      // cleaning thread in {@link azkaban.execapp.FlowRunnerManager}
+      // cleaning up the same project version when creating hardlink
+      synchronized (projectVersion) {
```

Comment:
In `getProjectVersion()`, it seems the projectVersion returned is a new object. Will the synchronization really work if they are synchronized on different objects here?

Reply:
it will only return new object only when projectId, version don't present in the hash map.
```   
 final ProjectVersion projectVersion;
    synchronized (this.installedProjects) {
      projectVersion = this.installedProjects
          .computeIfAbsent(new Pair<>(flow.getProjectId(), flow.getVersion()),
              k -> new ProjectVersion(flow.getProjectId(), flow.getVersion()));
    }
    return projectVersion;
 ```

Reply:
Was it subject to race conditions then? Could you describe the problem you tried to fix in more details?